### PR TITLE
Alex/enhance error responses

### PR DESCRIPTION
**Pull request by larzhole**
## Description
- Added common HTTP Error classes extending `http-errors` located in `/src/errors/`.
- Added middleware to handle errors, returning the correct HTTP response or a 500.

I had to install [express-async-handler](https://www.npmjs.com/package/express-async-handler) in order to wrap `async` methods in a way that lets us throw these errors.

**Issue Number(s)**:
#16 

## Issue(s):
I'm not sure how to use express-async-handler with the Role Controller yet

Not able to follow the suggested interface because typescript does not allow constructors on interfaces.
I ended up pulling the interface out completely and handle JSON creation in the middleware.

```
export interface HttpException {
    constructor(errors?: ValidationError[], message?: string)
    toJSON(): { message: string; errors?: ValidationError[] }
}
```

### CODE REVIEW

File: src/middleware/errorHandler.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      }
+
+      if (err.errors?.length > 0) {
+        json = { ...json, errors: err.errors }
+      }
+
+      return res.status(err.status).header('Content-Type', 'application/json; charset=utf-8').json(json)
+    }
+
+    return res.status(500).json({ message: 'An unexpected error has occurred. Please try again.' })
```
Comment (ericdowell) [inline src/middleware/errorHandler.ts:22]: You can reduce nesting above and instead check for inverse of instanceof HttpError. Then return early with this single line.
```ts
if (!(err instanceof HttpError)) {
  return res.status(500).json({ message: 'An unexpected error has occurred. Please try again.' })
}
```
Also making sure to use `safeStringify`, `header` and `end` with the `content` as `strong-error-handler` does in `send-json.js`.

File: src/errors/unprocessableEntity.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { ValidationError } from 'express-validator'
+import { UnprocessableEntity as HttpErrorsUnprocessableEntity } from 'http-errors'
+import { ErrorMessage } from './errorMessage'
+
+export class UnprocessableEntity extends HttpErrorsUnprocessableEntity {
+  errors: ValidationError[]
+  message: string
+
+  constructor(options?: { errors?: ValidationError[]; message?: string }) {
```
Comment (ericdowell) [inline src/errors/unprocessableEntity.ts:9]: We can make this into a custom type:
```ts
interface HttpExceptionOptions {
    errors?: ValidationError[];
    message?: string
}
```
Within `src/types/httpException.d.ts` and then use it for each constructor:
```ts
   constructor(options?: HttpExceptionOptions) {
```
That way it is in one place.

File: src/middleware/errorHandler.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { Request, Response, NextFunction } from 'express'
+import { HttpError } from 'http-errors'
+
+export default function (): any {
+  return async (err: Error, req: Request, res: Response, next: NextFunction): Promise<Response | void> => {
```
[the remark this answers is not in the document]
Comment (larzhole) [reply src/middleware/errorHandler.ts]: Complete

File: src/middleware/errorHandler.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    if (res.headersSent) {
+      return next(err)
+    }
+
+    if (err instanceof HttpError) {
+      let json: any = {
+        message: err.message,
+      }
+
+      if (err.errors?.length > 0) {
```
[the remark this answers is not in the document]
Comment (larzhole) [reply src/middleware/errorHandler.ts]: Complete

File: src/middleware/errorHandler.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { Request, Response, NextFunction } from 'express'
+import { HttpError } from 'http-errors'
+
+export default function (): any {
+  return async (err: Error, req: Request, res: Response, next: NextFunction): Promise<Response | void> => {
+    if (res.headersSent) {
+      return next(err)
```
[the remark this answers is not in the document]
Comment (larzhole) [reply src/middleware/errorHandler.ts]: Complete

File: src/middleware/errorHandler.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    if (err instanceof HttpError) {
+      let json: any = {
+        message: err.message,
+      }
+
+      if (err.errors?.length > 0) {
+        json = { ...json, errors: err.errors }
+      }
+
+      return res.status(err.status).header('Content-Type', 'application/json; charset=utf-8').json(json)
```
[the remark this answers is not in the document]
Comment (larzhole) [reply src/middleware/errorHandler.ts]: Complete

File: src/middleware/errorHandler.ts (hunk 1 above)
Comment (larzhole) [reply src/middleware/errorHandler.ts]: Complete

File: src/errors/unprocessableEntity.ts (hunk 2 above)
Comment (larzhole) [reply src/errors/unprocessableEntity.ts]: Complete

File: src/middleware/errorHandler.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+export default async (err: Error, req: Request, res: Response, next: NextFunction): Promise<Response | void> => {
+  if (res.headersSent) {
+    return req.socket.destroy()
+  }
+
+  if (!(err instanceof HttpError)) {
+    return res
+      .status(500)
+      .header('Content-Type', 'application/json; charset=utf-8')
+      .send({ message: 'An unexpected error has occurred. Please try again.' })
```
Comment (ericdowell) [inline src/middleware/errorHandler.ts:15]: This should use `safeStringify`:
```suggestion
      .send(safeStringify({ message: 'An unexpected error has occurred. Please try again.' }))
```

Comment (larzhole) [reply src/middleware/errorHandler.ts]: Complete

**Pull request merged**